### PR TITLE
Fix #9685 measurements issues and added some enhancements to it

### LINKED ISSUE #9685: Extend the GeoProcessing tool to work with client side Vector layers as WPS inputs
## Description

Our client is asking for a new additional feature based on the recently developed GeoProcessing tool for supported **Intersection** and **Buffer** processes. 
The existing tool functionalities are expecting to be used only for WMS and/or WFS layers in TOC as inputs for aforementioned processes.
As part of this work it is requested to extend the current tool to also support vector layers defined or imported on the client side.

Possible use case scenarios to supports are:

- Use the output of the GeoProcessing tool (vector layers generated by a buffer or intersection geoprocess) for performing another operation using it as a new input
- Use an imported vector layer (SHP, GeoJson, KML etc) as an input for the GeoProcessing tool
- Use existing annotations as an input for the GeoProcessing tool (to check if possible, it could be excluded initially since we have a new implementation in master)
- Use vector layers TOC coming from measurements

In all cases the UX should be the same and well consistent.

Partial evaluation already done in https://github.com/geosolutions-it/MapStore-C265/issues/29#issuecomment-1752913734

## Acceptance criteria

- [ ] Allow local vector layers listed above to be used as input in the GeoProcessing plugin
  - [ ] measurements
  - [ ] imported GeoJSON
  - [ ] imported Shapefiles
  - [ ] other imports
  - [ ] annotations
- [ ] make sure that the CollectGeometries request sent to the backend is adapted for vector layers
- [ ] handle cases where Geometry collection is generated from a vector layer, because buffer and intersect processes fails with it, we should prevent the user to click run and notifying it via user feedback
- [ ] for intersection it should maintain the previous checks we do, first layer/ feature can't be a point
- [ ] Add possibility to configure a predefined geoserver url to be used as a fallback within process requests when the source layer is a local vector layer

### PR DESCRIPTION
**Pull request by MV88**
## Description

- [ ] Fix vector measurement layer issue
- [ ] Filter out point text label as annotations 
- [ ] Rename features id to use measureType numbered with a counter
- [ ] Transform geodesic lines before calling the process and for highlight features

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

Fix #9685

**What is the new behavior?**

arc length as a buffer and blue line highlight
image

intersection of a buffer with states, (check lower parts should be the dark one 
image

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/reducers/geoProcessing.js [hunk 1]
```diff
@@ -181,7 +183,8 @@ function geoProcessing( state = {
             ...state,
             [action.source]: {
                 ...state[action.source],
-                features: (state[action.source].features || []).concat(action.data.features || []),
+                features: (state[action.source].features || []).concat(action.data.features || []).filter(f => !f?.properties?.measureId).map((ft, i)  => ({...ft,
+                    id: createFeatureId(ft, i)})),
```
Comment (offtherailz) [inline web/client/reducers/geoProcessing.js:187]: Please explain why you are filtering out all features that do not have `measureId` and creating ids for all features. It is not trivial the reason and it is  a special behaviour that have to be commented.
I suggest to do this changes outside the reducer. The reducer should expect well formed features.

File: web/client/utils/GeoProcessingUtils.js [hunk 2]
```diff
@@ -42,3 +45,30 @@ export const getCounter = (layers, groupName) => {
...
+ * it creates an id of a feature if not existing
+ * @param {object} feature list of layers to check
+ * @param {string} index the index of the feature
+ * @return {string} a predictable id
+ */
+export const createFeatureId = (feature, index) => {
+    return feature.id || `Feature #${index}`;
+};
+/**
+ * it creates an arc line into a multiple segments, only for LineString ft
```
Comment (offtherailz) [inline web/client/utils/GeoProcessingUtils.js:59]:
```suggestion
 * Transforms a line into an arc of multiple segments if `geodesic` property is equal to true.
```

File: web/client/utils/GeoProcessingUtils.js [hunk 3]
```diff
@@ -42,3 +45,30 @@ export const getCounter = (layers, groupName) => {
...
+ * @param {object} feature list of layers to check
+ * @return {object} the transformed feature if geodesic
```
Comment (offtherailz) [inline web/client/utils/GeoProcessingUtils.js:61]:
```suggestion
 * @return {object} the transformed feature if `properties.geodesic=true`, the original geometry in the other cases.
```

File: web/client/utils/GeoProcessingUtils.js [hunk 4]
```diff
@@ -42,3 +45,30 @@ export const getCounter = (layers, groupName) => {
...
+ */
+export const transformLineToArc = (feature) => {
+    if (feature?.properties?.geodesic) {
```
Comment (offtherailz) [inline web/client/utils/GeoProcessingUtils.js:64]: If it handles only linestrings, than do this check before transformation.

File: web/client/epics/geoProcessing.js [hunk 5]
```diff
@@ -589,12 +603,12 @@ export const runBufferProcessGPTEpic = (action$, store) => action$
                         }));
                     })
         );
-        const feature = sourceFeatureSelector(state);
+        const feature = transformLineToArc(sourceFeatureSelector(state));
```
Comment (offtherailz) [inline web/client/epics/geoProcessing.js:606]: The fact that we have to transform the geometry in 8 different points of the code risk to make the feature not maintainable. Try to centralize this behavior in a single place (selector).
Try to do also the normalization once (labels, ids, managemant of measureId etc...). The geodesic tool should be as much transparent as possible to the special case of measure).

File: web/client/utils/GeoProcessingUtils.js [hunk 6]
```diff
@@ -42,3 +45,30 @@ export const getCounter = (layers, groupName) => {
...
+ */
+export const createFeatureId = (feature, index) => {
+    return feature.id || `Feature #${index}`;
+};
+/**
+ * it creates an arc line into a multiple segments, only for LineString ft
+ * @param {object} feature list of layers to check
+ * @return {object} the transformed feature if geodesic
+ */
+export const transformLineToArc = (feature) => {
```
Comment (offtherailz) [inline web/client/utils/GeoProcessingUtils.js:63]: I'd name it `densifyGeodesicFeatures`

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [ ] See comments inline
- [ ] When I try to select, I have an error and a success notification at the same time.
[screencast-github.com-2023.12.13-10_35_57.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/0aec8a41-443a-41d7-8ae3-f78cfa7d2b60)

- [ ] Selection do not look to work as expected. 
- [ ] When I try to select the buffered geometry, I have a crash.
[screencast-github.com-2023.12.13-10_37_09.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/91739d5a-a51e-4a91-971a-90b5fc7b1de6)

Error in console:
```
StandardRouter.jsx:86 TypeError: Cannot read properties of undefined (reading 'id')
    at createFeatureId (GeoProcessingUtils.js:77:18)
    at eval (SourceLayer.jsx:119:89)
    at Array.map (<anonymous>)
    at Source (SourceLayer.jsx:118:29)
    at renderWithHooks (react-dom.development.js:16320:18)
    at updateFunctionComponent (react-dom.development.js:18284:20)
    at beginWork$1 (react-dom.development.js:20098:16)
    at HTMLUnknownElement.callCallback (react-dom.development.js:362:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:411:16)
    at invokeGuardedCallback (react-dom.development.js:466:31)
    at beginWork$$1 (react-dom.development.js:25730:7)
    at performUnitOfWork (react-dom.development.js:24638:12)
    at workLoopSync (react-dom.development.js:24614:22)
    at performSyncWorkOnRoot (react-dom.development.js:24182:11)
    at eval (react-dom.development.js:12238:24)
    at unstable_runWithPriority (scheduler.development.js:815:12) '\n    at ErrorBoundary (webpack://mapstore2/./node_modules/react-error-boundary/dist/commonjs/ErrorBoundary.js?:36:5)\n    at div\n    at Router (webpack://mapstore2/./node_modules/react-router/es/Router.js?:36:5)\n    at ConnectedRouter (webpack://mapstore2/./node_modules/connected-react-router/esm/ConnectedRouter.js?:60:7)\n    at ConnectedRouterWithContext (webpack://mapstore2/./node_modules/connected-react-router/esm/ConnectedRouter.js?:158:25)\n    at Connect (webpack://mapstore2/./node_modules/react-redux/es/components/connectAdvanced.js?:152:37)\n    at IntlProvider (webpack://mapstore2/./node_modules/react-intl/lib/index.es.js?:804:9)\n    at Localized (webpack://mapstore2/./web/client/components/I18N/Localized.jsx?:28:5)\n    at eval (webpack://mapstore2/./web/client/components/theme/Theme.jsx?:42:5)\n    at Connect (webpack://mapstore2/./node_modules/react-redux/es/components/connectAdvanced.js?:152:37)\n    at div\n    at StandardRouter (webpack://mapstore2/./web/client/components/app/StandardRouter.jsx?:51:5)\n    at Connect (webpack://mapstore2/./node_modules/react-redux/es/components/connectAdvanced.js?:152:37)\n    at DragDropContextContainer (webpack://mapstore2/./node_modules/react-dnd/lib/DragDropContext.js?:81:5)\n    at Provider (webpack://mapstore2/./node_modules/react-redux/es/components/Provider.js?:24:24)\n    at StandardApp (webpack://mapstore2/./web/client/components/app/StandardApp.jsx?:78:5)\n    at WithExtensions (webpack://mapstore2/./web/client/components/app/withExtensions.js?:50:7)'
```

File: web/client/utils/GeoProcessingUtils.js (hunk 4 above)
Comment (MV88) [reply web/client/utils/GeoProcessingUtils.js:64]: what about adding a condition check for the geometry property? doing it external to the function is worse imo

File: web/client/epics/geoProcessing.js (hunk 5 above)
Comment (MV88) [reply web/client/epics/geoProcessing.js:606]: sounds good, let me better see this

File: web/client/reducers/geoProcessing.js (hunk 1 above)
Comment (MV88) [reply web/client/reducers/geoProcessing.js:187]: the main problem was that the vector layer generated from feature measurements are missing ids and it makes some other part of the code to fail. i'll centralize this behaviour in the layers selector of the geoprocessing tool and also in epics

**Comment by MV88** [comment]
@offtherailz this pr is ready for a new review. 
i've tested many use cases as you can see here

https://github.com/geosolutions-it/MapStore2/assets/11991428/9687df3a-b19b-4cf5-923f-a8f460942c73

**Review by offtherailz — CHANGES_REQUESTED** [review]
I'll test it as well as the changes suggested are applied. 
Thank you .

File: web/client/selectors/geoProcessing.js [hunk 7]
```diff
@@ -73,5 +74,20 @@ export const wfsBackedLayersSelector = (state) => {
     const layers = layersSelector(state);
     return layers
         .filter(l => l.group !== "background")
-        .filter(layer => hasWFSService(layer) || layer.type === "vector");
+        .filter(layer => hasWFSService(layer) || layer.type === "vector")
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/selectors/geoProcessing.js:77]: but it does not fetch only vector layers

**Comment by MV88** [comment]
@offtherailz review is ready

File: web/client/epics/widgets.js [hunk 8]
```diff
@@ -199,7 +199,7 @@ export const toggleWidgetConnectFlow = (action$, {getState = () => {}} = {}) =>
                             if (widget.widgetType === 'map') {
                                 deps = deps.filter(d => (WIDGETS_MAPS_REGEX.exec(d) || [])[2] === widget.selectedMapId);
                             }
-                            return configureDependency(active, deps[0], options, widget.dependeciesMap).concat(Rx.Observable.of(toggleDependencySelector(false, {})));
+                            return configureDependency(active, deps[0], options, widget.dependenciesMap).concat(Rx.Observable.of(toggleDependencySelector(false, {})));
```
Comment (offtherailz) [inline web/client/epics/widgets.js:202]: I think this is a fix for another functionality. @dsuren1 is working on widget's fixes too, tell to him this and remove from this PR.

**Review by offtherailz — CHANGES_REQUESTED** [review]
I'm trying to use this map to select the buffer layer but I see a warning of no feature clicked.

[geoprocessing_map.json](https://github.com/geosolutions-it/MapStore2/files/13683440/geoprocessing_map.json)

[screencast-localhost_8081-2023.12.15-10_22_56.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/69013d91-35b5-4ed5-ac39-630516099b67)

same issue if I try to select as vector layer of input natural earth 10m admin data (Can not upload it here).

File: web/client/epics/widgets.js (hunk 8 above)
Comment (MV88) [reply web/client/epics/widgets.js:202]: correct

Comment (dsuren1) [reply web/client/epics/widgets.js:202]: The argument is not used by the function. Removed it in the widget fix PR

Comment (offtherailz) [reply web/client/epics/widgets.js:202]: So please remove this fix from your PR too @MV88 , so we can merge your contributions easely.

**Review by offtherailz — APPROVED** [review]
:+1:  works quite fine. 
The only things I noticed here: 
- Circle annotation cause an error because internally is handled as a point
- When using big sets of data (e.g. natural earth 10m admin boundaries) the geoprocessing tool slows down a bit. It happens also to the map, until I hide the layer anyway. Probably some optimizations can be done on this side.

Because circle annotation was not strictly a part of this issue and because the performance are anyway acceptable under stress, I think I can approve this PR, that fulfills the requirements and does not introduce any regression.

Please @MV88 and @tdipisa decide what to do with my notes. 

Thank you for your development @MV88 now this tool is really powerful.

**Comment by tdipisa** [comment]
@offtherailz thank you

> When using big sets of data (e.g. natural earth 10m admin boundaries) the geoprocessing tool slows down a bit. It happens also to the map, until I hide the layer anyway. Probably some optimizations can be done on this side.

it would be good to understand where is the performance issue, if it is on the client side when rendering complex features or on the WPS side.

**Review by offtherailz — APPROVED** [review]
:rocket:  :+1: 
image

**Comment by tdipisa** [comment]
@ElenaGallo please test in DEV once deployed.

**Comment by ElenaGallo** [comment]
Test passed @MV88 please backport to 2023.02.xx. Thanks

**Comment by MV88** [comment]
@offtherailz backport is here https://github.com/geosolutions-it/MapStore2/pull/9835 for you to review

**Pull request merged**